reactor-core/src/main/java/reactor/core/publisher/FluxOnAssembly.java

```diff
@@ -298,37 +443,123 @@ public String getMessage() {
+						maxOperatorWidth = length;
+					}
+
+					final DebugStats stats = t.debugStats;
+					if (stats != null) {
+						final long requested = stats.requested();
+						length = requested == Long.MAX_VALUE ? "∞".length() : String.valueOf(requested).length();
+
+						if (length > maxRequestedWidth) {
+							maxRequestedWidth = length;
+						}
+
+						length = String.valueOf(stats.produced()).length();
+
+						if (length > maxProducedWidth) {
+							maxProducedWidth = length;
+						}
 					}
 				}
 
 				StringBuilder sb = new StringBuilder(super.getMessage())
 						.append("\nError has been observed at the following site(s):\n");
-				for(Tuple4<Integer, String, String, Integer> t : chainOrder) {
-					Integer indent = t.getT4();
-					String operator = t.getT2();
-					String message = t.getT3();
+				for(Entry t : chainOrder) {
+					int indent = t.indent;
+					String operator = t.operator;
+					String message = t.line;
 					sb.append("\t|_");
 					for (int i = 0; i < indent; i++) {
 						sb.append("____");
 					}
 
-					for (int i = operator.length(); i < maxWidth + 1; i++) {
+					for (int i = operator.length(); i < maxOperatorWidth + 1; i++) {
 						sb.append(' ');
 					}
 					sb.append(operator);
 					sb.append(Traces.CALL_SITE_GLUE);
 					sb.append(message);
+					final DebugStats stats = t.debugStats;
+					if (stats != null) {
+						sb.append("\n");
+						for (int i = 0; i < indent; i++) {
+							sb.append("    ");
+						}
+						for (int i = 0; i < maxOperatorWidth + 1; i++) {
+							sb.append(' ');
+						}
+
+						sb.append("requested: ");
+						String requested = stats.requested() == Long.MAX_VALUE ? "∞" : String.valueOf(stats.requested());
+						for (int i = requested.length(); i < maxRequestedWidth + 1; i++) {
+							sb.append(' ');
+						}
+						sb.append(requested);
+						sb.append("; ");
+
+						sb.append("produced: ");
+						String produced = String.valueOf(stats.produced());
+						for (int i = produced.length(); i < maxProducedWidth + 1; i++) {
+							sb.append(' ');
+						}
+						sb.append(produced);
+						sb.append("; ");
+
+						sb.append("lastRequestAt: ");
+						sb.append(DateTimeFormatter.ISO_INSTANT.format(stats.lastRequestTime()));
+						sb.append("; ");
+
+						sb.append("lastOnNextAt: ");
+						sb.append(DateTimeFormatter.ISO_INSTANT.format(stats.lastOnNextTime()));
+						sb.append("; ");
+
+						sb.append("state: ");
+
+						if (stats.isCompleted()) {
+							sb.append("completed");
+						} else if (stats.isErrored()) {
+							sb.append("  errored");
+						} else if (stats.isCancelled()) {
+							sb.append("cancelled");
+						} else {
+							sb.append("  running");
+						}
+					}
 					sb.append("\n");
 				}
 				sb.append("Stack trace:");
 				return sb.toString();
 			}
 		}
+
+		static class Entry {
+
+			final int        key;
+			final int        indent;
+			final String     operator;
+			final String     line;
+			@Nullable
```

Comment:
What about the idea of No-op `DebugStats` that we discussed?

Reply:
It can be challenging to then check if the DebugStats are real or noops when it comes to the printing of the message (and in the end it will end up to check `if debugStats instanceof NoOpsDebugStats`), thus I decided to stick to null check since that logic is already present in the message printing and JIT should eliminate such checks at runtime in any case